yb-voyager/src/srcdb/postgres.go

```diff
@@ -104,6 +112,502 @@ func (pg *PostgreSQL) Disconnect() {
+	}
+
+	// Check if tables have SELECT permission
+	missingTables, tablesWithNoUsagePerm, err := pg.checkWhichTablesDontHaveSelectPermission()
+	if err != nil {
+		return nil, fmt.Errorf("error checking table select permissions: %w", err)
+	}
+	if len(missingTables) > 0 {
+		combinedResult = append(combinedResult, fmt.Sprintf("Tables missing SELECT permission: %v", missingTables))
+	}
+	if len(tablesWithNoUsagePerm) > 0 {
+		// Print that some tables could not be checked for SELECT permission as the parent schema does not have USAGE permission
+		combinedResult = append(combinedResult, fmt.Sprintf("Following tables could not be checked for SELECT permission as the parent schema does not have USAGE permission: %v", tablesWithNoUsagePerm))
+	}
+
+	// Return combined result of checks if any issues, else return nothing (empty string and nil)
+	return combinedResult, nil
+
+}
+
+func (pg *PostgreSQL) checkWhichTablesDontHaveSelectPermission() (tablesWithMissingPerm []string, tablesWithNoUsagePerm []string, err error) {
+	// Users only need SELECT permissions on the tables of the schema they want to export for export schema
+	trimmedSchemaList := pg.getTrimmedSchemaList()
+	trimmedSchemaList = append(trimmedSchemaList, "pg_catalog", "information_schema")
+	querySchemaList := "'" + strings.Join(trimmedSchemaList, "','") + "'"
+
+	checkTableSelectPermissionQuery := fmt.Sprintf(`WITH schema_list AS (
+		SELECT unnest(ARRAY[%s]) AS schema_name
+	),
+	accessible_schemas AS (
+		SELECT schema_name
+		FROM schema_list
+		WHERE has_schema_privilege('%s', quote_ident(schema_name), 'USAGE')
+	)
+	SELECT
+		t.schemaname AS schema_name,
+		t.tablename AS table_name,
+		CASE WHEN has_table_privilege('%s', quote_ident(t.schemaname) || '.' || quote_ident(t.tablename), 'SELECT') THEN 'Granted' ELSE 'Missing' END AS status
+	FROM pg_tables t
+	JOIN accessible_schemas a ON t.schemaname = a.schema_name
+	UNION ALL
+	SELECT
+		t.schemaname AS schema_name,
+		t.tablename AS table_name,
+		'No Usage Permission On The Table Parent Schema' AS status
+	FROM pg_tables t
+	WHERE t.schemaname IN (SELECT schema_name FROM schema_list)
+	AND NOT EXISTS (
+		SELECT 1
+		FROM accessible_schemas a
+		WHERE t.schemaname = a.schema_name
+	);`, querySchemaList, pg.source.User, pg.source.User)
+	rows, err := pg.db.Query(checkTableSelectPermissionQuery)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error in querying(%q) source database for checking table select permission: %w", checkTableSelectPermissionQuery, err)
+	}
+	defer func() {
+		closeErr := rows.Close()
+		if closeErr != nil {
+			log.Warnf("close rows for query %q: %v", checkTableSelectPermissionQuery, closeErr)
+		}
+	}()
+
+	// If result is No Usage Permission On The Table Parent Schema, then the schema itself doesn't have USAGE permission store them in tablesWithNoUsagePerm
+	var tableSchemaName, tableName, status string
```

Comment:
Yeah. But some queries return different values. Like they tell what exactly is missing. I have intentionally kept them that way for now so that we can change their usage easily. But let's discuss about this

Reply:
Yeah made the change